Gutter compat

dabbler0 opened the pull request:
Add methods to Droplet to mimic Ace editor gutter decoration functionality. Changes from existing behavior:
 - Droplet now has `setBreakpoint`, `clearBreakpoint`, `clearBreakpoints`, `getBreakpoints`, and `setAnnotations`, with behavior identical to Ace. Droplet will set all of these for itself when called and will also set them in its corresponding Ace editor.
 - Droplet's `guttermousemove` now also fires when Ace's fires
 - Droplet annotation texts can be read with a title-text
 - Droplet line numbers are now on top and can recieve mouseover events directly
 - Droplet annotation glyphs are centered vertically on lines

cpirich commented on css/droplet.css:
```diff
@@ -256,3 +256,14 @@
+.droplet_error {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAMAAAAoLQ9TAAABOFBMVEX/////////QRswFAb/Ui4wFAYwFAYwFAaWGAfDRymzOSH/PxswFAb/SiUwFAYwFAbUPRvjQiDllog5HhHdRybsTi3/Tyv9Tir+Syj/UC3////XurebMBIwFAb/RSHbPx/gUzfdwL3kzMivKBAwFAbbvbnhPx66NhowFAYwFAaZJg8wFAaxKBDZurf/RB6mMxb/SCMwFAYwFAbxQB3+RB4wFAb/Qhy4Oh+4QifbNRcwFAYwFAYwFAb/QRzdNhgwFAYwFAbav7v/Uy7oaE68MBK5LxLewr/r2NXewLswFAax ...
+  background-repeat: no-repeat;
+  background-position: 2px 35%;
```
The 35% thing was a bit of a hack I put in to deal with placing it in a container that already had the padding-top set, but not the corresponding padding-bottom. If you find a way to make this "2px center", it would be more reliable (and make more sense)

cpirich commented on css/droplet.css:
```diff
@@ -256,3 +256,14 @@
+.droplet_warning {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAMAAAAoLQ9TAAAAmVBMVEX///8AAAD///8AAAAAAABPSzb/5sAAAAB/blH/73z/ulkAAAAAAAD85pkAAAAAAAACAgP/vGz/rkDerGbGrV7/pkQICAf////e0IsAAAD/oED/qTvhrnUAAAD/yHD/njcAAADuv2r/nz//oTj/p064oGf/zHAAAAA9Nir/tFIAAAD/tlTiuWf/tkIAAACynXEAAAAAAAAtIRW7zBpBAAAAM3RSTlMAABR1m7RXO8Ln31Z36zT+neXe5OzooRDfn+TZ4p3h2hTf4t3k3ucyrN1K5+Xaks52Sfs9CXgrAAAA ...
+  background-repeat: no-repeat;
+  background-position: 2px 35%;
```
same as above

cpirich commented on src/controller.coffee:
```diff
@@ -4226,54 +4227,82 @@ hook 'populate', 0, ->
+    @annotations[el.row] ?= []
+    @annotations[el.row].push el
+
+  console.log 'Just set annotations. They are now', @annotations
```
remove console.log

cpirich commented on src/controller.coffee:
```diff
@@ -4287,15 +4316,31 @@ Editor::addLineNumberForLine = (line) ->
+  # and graphics
+  if @annotations[line]?
+    lineDiv.className += ' droplet_' + getMostSevereAnnotationType(@annotations[line])
+    lineDiv.style.backgroundPosition = "2px #{treeView.distanceToBase[line].above - @view.opts.textHeight - @fontAscent}px"
```
is this taking precedence over the earlier CSS? If you find a way to make it work with "2px center" in the CSS, you may not need to do this.

cpirich commented on src/controller.coffee:
```diff
@@ -4287,15 +4316,31 @@ Editor::addLineNumberForLine = (line) ->
+    lineDiv.className += ' droplet_breakpoint'
+
+  lineDiv.style.top = "#{treeView.bounds[line].y}px"
   lineDiv.style.paddingTop = "#{treeView.distanceToBase[line].above - @view.opts.textHeight - @fontAscent}px"
```
This paddingTop line is the reason why using center is difficult. It shifts the rectangle down. Perhaps both top and bottom padding can be set, preserving the vertical centering?

cpirich commented on src/controller.coffee:
```diff
@@ -4287,15 +4316,31 @@ Editor::addLineNumberForLine = (line) ->
+  for el, i in arr
+    if el.type is 'error'
+      return 'error'
+  return 'warning'
```
Are annotations guaranteed to always be 'error' or 'warning'? (I'm not familiar with the spec)

cpirich commented:
Looks good - outside of the minor issues I commented on.

dabbler0 commented:
Responding to earlier comments:
 - Ace editor supports only types `info`, `error`, and `warning`, so I've covered those now.
 - I've now changed the css to `2px center` and added a paddingBottom, instead of computing the `backgroundPosition` on-the-fly.

cpirich commented:
LGTM!
